### PR TITLE
2615 feature add query to savesets

### PR DESCRIPTION
**Pull request by evans-g-crsj**

### CODE REVIEW

File: modules/mapping-utils/src/resolveSets.js [hunk 1]
```diff
@@ -149,48 +175,193 @@ export const deleteSaveSets = ({ callback }) => async (
...
-        lang: 'painless',
-        source: `ctx._source['tag'] = '${tag}'`,
-      },
-      query: {
-        bool: {
-          filter: {
-            term: { userId: userId },
+  const { sourceType } = source;
+
+  switch (subAction) {
```
Comment (francisl) [inline modules/mapping-utils/src/resolveSets.js]: this switch case is way too big. pls split it in logic blocks, inner function and then a clean switch statement.

File: modules/mapping-utils/src/resolveSets.js [hunk 2]
```diff
@@ -149,48 +175,193 @@ export const deleteSaveSets = ({ callback }) => async (
...
+    case SubActionTypes.REMOVE_IDS:
+    case SubActionTypes.ADD_IDS: {
+      if (SourceType.QUERY === sourceType) {
+        const { type, sqon, path } = data;
+
+        if (isQueryEmpty(sqon)) {
```
Comment (francisl) [inline modules/mapping-utils/src/resolveSets.js]: I don't see how a empty query is a exception. This a a completly valid state, maybe we can't work on it and need to do something, but still a valid state, it's not 'exceptional'

maybe there is something I don't yet understand with arranger. But throwing here and there for business logic a la java, is IMO, a bad pratice. We should be clear about the flow.

Plus, the exception seem to be unhandled later on.

https://dzone.com/articles/exceptions-as-controlflow-in-java#:~:text=One%20of%20these%20common%20bad,makes%20your%20code%20less%20readable.&text=For%20this%20reason%2C%20the%20functional%20programming%20paradigm%20does%20not%20welcome%20exceptions.

https://reflectoring.io/business-exceptions/

File: modules/mapping-utils/src/resolveSets.js [hunk 3]
```diff
@@ -149,48 +175,193 @@ export const deleteSaveSets = ({ callback }) => async (
...
           },
-          must: {
-            term: {
-              setId: {
-                value: setId,
+        });
+
+        const sets = mapHits(esSearchResponse);
+        if (sets.length === 0) {
+          throw new Error('Internal error');
```
Comment (francisl) [inline modules/mapping-utils/src/resolveSets.js]: why having no results is and exception? does it mean the system crashed? Somebody cut a wired?

File: modules/mapping-utils/src/resolveSets.js [hunk 4]
```diff
@@ -149,48 +175,193 @@ export const deleteSaveSets = ({ callback }) => async (
...
+          es,
+          index: index,
+          type: es_type,
+          query,
+          path,
+          sort: [{ field: '_id', order: 'asc' }],
+        });
+
+        if (idsFromQuery.length === 0) {
+          return;
```
Comment (francisl) [inline modules/mapping-utils/src/resolveSets.js]: why undefined is fine here? (following the exception handling)

**Review by francisl — CHANGES_REQUESTED** [review]
overall good.

I only have a disagrement about the use of exception for as control flow. Please take a look, then we can discuss and see how we proceed.

Thanks

File: modules/mapping-utils/src/resolveSets.js (hunk 1 above)
Comment (evans-g-crsj) [reply modules/mapping-utils/src/resolveSets.js]: Good point

File: modules/mapping-utils/src/resolveSets.js (hunk 2 above)
Comment (evans-g-crsj) [reply modules/mapping-utils/src/resolveSets.js]: Absolutely. Thanks

File: modules/mapping-utils/src/resolveSets.js (hunk 3 above)
Comment (evans-g-crsj) [reply modules/mapping-utils/src/resolveSets.js]: haha good point

File: modules/mapping-utils/src/resolveSets.js [hunk 5]
```diff
@@ -67,17 +269,18 @@ export const saveSet = ({ types, callback }) => async (
   if (tag) {
     // if a tag is present, test early.
     if (!isTagValid(tag)) {
-      throw new Error('Invalid tag, no set created.');
+      return null;
     } else if (!isFunction(callback)) {
-      throw new Error('Cannot process further, no set created.');
+      return null;
```
Comment (adipaul1981) [inline modules/mapping-utils/src/resolveSets.js:272]: since return is the same. cant they be merged into the same if condition?

**Review by adipaul1981 — APPROVED** [review]

File: modules/mapping-utils/src/resolveSets.js (hunk 5 above)
Comment (evans-g-crsj) [reply modules/mapping-utils/src/resolveSets.js:272]: good point

**Pull request merged**